// @midwayjs/core/dist/common/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALUE_TYPE = exports.KEYS = exports.FUNCTION_INJECT_KEY = void 0;
exports.FUNCTION_INJECT_KEY = 'midway:function_inject_key';
/**
 * 静态参数
 *
 */
exports.KEYS = {
    OBJECTS_ELEMENT: 'objects',
    OBJECT_ELEMENT: 'object',
    IMPORT_ELEMENT: 'import',
    PROPERTY_ELEMENT: 'property',
    LIST_ELEMENT: 'list',
    MAP_ELEMENT: 'map',
    ENTRY_ELEMENT: 'entry',
    VALUE_ELEMENT: 'value',
    PROPS_ELEMENT: 'props',
    PROP_ELEMENT: 'prop',
    SET_ELEMENT: 'set',
    CONSTRUCTORARG_ELEMENT: 'constructor-arg',
    REF_ELEMENT: 'ref',
    JSON_ELEMENT: 'json',
    CONFIGURATION_ELEMENT: 'configuration',
    ID_ATTRIBUTE: 'id',
    PATH_ATTRIBUTE: 'path',
    DIRECT_ATTRIBUTE: 'direct',
    AUTOWIRE_ATTRIBUTE: 'autowire',
    ASYNC_ATTRIBUTE: 'async',
    NAME_ATTRIBUTE: 'name',
    REF_ATTRIBUTE: 'ref',
    KEY_ATTRIBUTE: 'key',
    VALUE_ATTRIBUTE: 'value',
    TYPE_ATTRIBUTE: 'type',
    EXTERNAL_ATTRIBUTE: 'external',
    OBJECT_ATTRIBUTE: 'object',
    RESOURCE_ATTRIBUTE: 'resource',
    SCOPE_ATTRIBUTE: 'scope',
    ASPECT_ELEMENT: 'aspect',
    AROUND_ELEMENT: 'around',
    EXPRESSION_ATTRIBUTE: 'expression',
    EXECUTE_ATTRIBUTE: 'execute',
};
exports.VALUE_TYPE = {
    STRING: 'string',
    DATE: 'date',
    NUMBER: 'number',
    INTEGER: 'int',
    TEMPLATE: 'template',
    MANAGED: 'managed',
    OBJECT: 'object',
    BOOLEAN: 'boolean',
};
