// @midwayjs/upload/dist/middleware.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UploadMiddleware = void 0;
const decorator_1 = require("@midwayjs/decorator");
const path_1 = require("path");
const fs_1 = require("fs");
const stream_1 = require("stream");
const _1 = require(".");
const parse_1 = require("./parse");
const getRawBody = require("raw-body");
const { unlink, writeFile } = fs_1.promises;
let UploadMiddleware = class UploadMiddleware {
    resolve(app) {
        if (app.getFrameworkType() === decorator_1.MidwayFrameworkType.WEB_EXPRESS) {
            return async (req, res, next) => {
                return this.execUpload(req, req, res, next, true);
            };
        }
        else {
            return async (ctx, next) => {
                var _a;
                const req = ((_a = ctx.request) === null || _a === void 0 ? void 0 : _a.req) || ctx.request;
                return this.execUpload(ctx, req, ctx, next, false);
            };
        }
    }
    async execUpload(ctx, req, res, next, isExpress) {
        const { mode, tmpdir, fileSize } = this.upload;
        const boundary = this.getUploadBoundary(req);
        if (!boundary) {
            return next();
        }
        ctx.fields = {};
        ctx.files = [];
        ctx.cleanupRequestFiles = async () => {
            var _a;
            if (!((_a = ctx.files) === null || _a === void 0 ? void 0 : _a.length)) {
                return [];
            }
            return Promise.all(ctx.files.map(async (fileInfo) => {
                if (typeof fileInfo.data !== 'string') {
                    return false;
                }
                try {
                    await unlink(fileInfo.data);
                    return true;
                }
                catch (_a) {
                    return false;
                }
            }));
        };
        let body;
        if (this.isReadableStream(req, isExpress)) {
            if (mode === 'stream') {
                const { fields, fileInfo } = await (0, parse_1.parseFromReadableStream)(req, boundary);
                if (!this.checkExt(fileInfo.filename)) {
                    throw new _1.MultipartInvalidFilenameError(fileInfo.filename);
                }
                else {
                    ctx.fields = fields;
                    ctx.files = [fileInfo];
                    return next();
                }
            }
            body = await getRawBody(req, {
                limit: fileSize,
            });
        }
        else {
            body = req.body;
        }
        const data = await (0, parse_1.parseMultipart)(body, boundary);
        if (!data) {
            return next();
        }
        ctx.fields = data.fields;
        const requireId = `upload_${Date.now()}.${Math.random()}`;
        const files = data.files;
        const notCheckFile = files.find(fileInfo => {
            if (!this.checkExt(fileInfo.filename)) {
                return fileInfo;
            }
        });
        if (notCheckFile) {
            throw new _1.MultipartInvalidFilenameError(notCheckFile.filename);
        }
        ctx.files = await Promise.all(files.map(async (file, index) => {
            const { data, filename } = file;
            if (mode === 'file') {
                const ext = (0, path_1.extname)(filename);
                const tmpFileName = (0, path_1.resolve)(tmpdir, `${requireId}.${index}${ext}`);
                await writeFile(tmpFileName, data, 'binary');
                file.data = tmpFileName;
            }
            else if (mode === 'stream') {
                file.data = new stream_1.Readable({
                    read() {
                        this.push(data);
                        this.push(null);
                    },
                });
            }
            return file;
        }));
        return next();
    }
    getUploadBoundary(request) {
        var _a;
        const method = (request.method || request.httpMethod || '').toUpperCase();
        if (method !== 'POST' &&
            method !== 'PUT' &&
            method !== 'DELETE' &&
            method !== 'PATCH') {
            return false;
        }
        if (!((_a = request.headers) === null || _a === void 0 ? void 0 : _a['content-type'])) {
            return false;
        }
        const contentType = request.headers['content-type'];
        if (!contentType.startsWith('multipart/form-data;')) {
            return false;
        }
        const boundaryMatch = /boundary=(.*)(;|\s|$)/.exec(contentType);
        if (!(boundaryMatch === null || boundaryMatch === void 0 ? void 0 : boundaryMatch[1])) {
            return false;
        }
        return boundaryMatch[1];
    }
    isReadableStream(req, isExpress) {
        // ref: https://github.com/rvagg/isstream/blob/master/isstream.js#L10
        if (req instanceof stream_1.Stream &&
            typeof req._read === 'function' &&
            typeof req._readableState === 'object' &&
            (!req.body || isExpress)) {
            return true;
        }
        if (req.pipe && req.on && !req.body) {
            return true;
        }
        return false;
    }
    checkExt(filename) {
        const ext = (0, path_1.extname)(filename).toLowerCase();
        const { whitelist } = this.upload;
        if (!Array.isArray(whitelist)) {
            return true;
        }
        return whitelist.includes(ext);
    }
    static getName() {
        return 'upload';
    }
};
__decorate([
    (0, decorator_1.Config)('upload'),
    __metadata("design:type", Object)
], UploadMiddleware.prototype, "upload", void 0);
__decorate([
    (0, decorator_1.Logger)(),
    __metadata("design:type", Object)
], UploadMiddleware.prototype, "logger", void 0);
UploadMiddleware = __decorate([
    (0, decorator_1.Middleware)()
], UploadMiddleware);
exports.UploadMiddleware = UploadMiddleware;
