// @midwayjs/core/dist/service/loggerService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MidwayLoggerService = void 0;
const decorator_1 = require("@midwayjs/decorator");
const configService_1 = require("./configService");
const serviceFactory_1 = require("../common/serviceFactory");
const logger_1 = require("@midwayjs/logger");
let MidwayLoggerService = class MidwayLoggerService extends serviceFactory_1.ServiceFactory {
    constructor(applicationContext) {
        super();
        this.applicationContext = applicationContext;
    }
    init() {
        var _a;
        this.initClients(this.configService.getConfiguration('midwayLogger'));
        // alias inject logger
        (_a = this.applicationContext) === null || _a === void 0 ? void 0 : _a.registerObject('logger', this.getLogger('appLogger'));
    }
    createClient(config, name) {
        logger_1.loggers.createLogger(name, config);
    }
    getName() {
        return 'logger';
    }
    createLogger(name, config) {
        return logger_1.loggers.createLogger(name, config);
    }
    getLogger(name) {
        return logger_1.loggers.getLogger(name);
    }
};
__decorate([
    (0, decorator_1.Inject)(),
    __metadata("design:type", configService_1.MidwayConfigService)
], MidwayLoggerService.prototype, "configService", void 0);
__decorate([
    (0, decorator_1.Init)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], MidwayLoggerService.prototype, "init", null);
MidwayLoggerService = __decorate([
    (0, decorator_1.Provide)(),
    (0, decorator_1.Scope)(decorator_1.ScopeEnum.Singleton),
    __metadata("design:paramtypes", [Object])
], MidwayLoggerService);
exports.MidwayLoggerService = MidwayLoggerService;
