// @midwayjs/mwcc/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCodeOfNode = exports.Analyzer = exports.template = exports.mergeConfigs = exports.resolveTsConfigFile = exports.getDefaultConfig = exports.Program = exports.CompilerHost = exports.ts = exports.tsquery = void 0;
const ts = __importStar(require("typescript"));
exports.ts = ts;
const query_1 = __importDefault(require("./tsquery/query"));
exports.tsquery = query_1.default;
__exportStar(require("./orchestra"), exports);
var compiler_host_1 = require("./compiler-host");
Object.defineProperty(exports, "CompilerHost", { enumerable: true, get: function () { return compiler_host_1.CompilerHost; } });
var program_1 = require("./program");
Object.defineProperty(exports, "Program", { enumerable: true, get: function () { return program_1.Program; } });
var config_1 = require("./config");
Object.defineProperty(exports, "getDefaultConfig", { enumerable: true, get: function () { return config_1.getDefaultConfig; } });
Object.defineProperty(exports, "resolveTsConfigFile", { enumerable: true, get: function () { return config_1.resolveTsConfigFile; } });
Object.defineProperty(exports, "mergeConfigs", { enumerable: true, get: function () { return config_1.mergeConfigs; } });
__exportStar(require("./type"), exports);
__exportStar(require("./transformation/type"), exports);
var template_1 = require("./template");
Object.defineProperty(exports, "template", { enumerable: true, get: function () { return template_1.template; } });
var analyze_1 = require("./comprehension/analyze");
Object.defineProperty(exports, "Analyzer", { enumerable: true, get: function () { return analyze_1.Analyzer; } });
var util_1 = require("./util");
Object.defineProperty(exports, "getCodeOfNode", { enumerable: true, get: function () { return util_1.getCodeOfNode; } });
