// @midwayjs/decorator/dist/decorator/framework/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationContext = exports.Logger = exports.App = exports.Config = exports.Plugin = void 0;
const __1 = require("../../");
function Plugin(identifier) {
    return (0, __1.createCustomPropertyDecorator)(__1.PLUGIN_KEY, {
        identifier,
    });
}
exports.Plugin = Plugin;
function Config(identifier) {
    return (0, __1.createCustomPropertyDecorator)(__1.CONFIG_KEY, {
        identifier,
    });
}
exports.Config = Config;
function App(typeOrNamespace) {
    return (0, __1.createCustomPropertyDecorator)(__1.APPLICATION_KEY, {
        type: typeOrNamespace,
    });
}
exports.App = App;
function Logger(identifier) {
    return (0, __1.createCustomPropertyDecorator)(__1.LOGGER_KEY, {
        identifier,
    });
}
exports.Logger = Logger;
function ApplicationContext() {
    return (0, __1.createCustomPropertyDecorator)(__1.APPLICATION_CONTEXT_KEY, {});
}
exports.ApplicationContext = ApplicationContext;
