// @cool-midway/core/configuration.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoolConfiguration = void 0;
const data_1 = require("./tag/data");
const config_1 = require("./module/config");
const index_1 = require("./event/index");
const func_1 = require("./util/func");
const filter_1 = require("./exception/filter");
const decorator_1 = require("@midwayjs/decorator");
const DefaultConfig = require("./config/config.default");
const koa = require("@midwayjs/koa");
const location_1 = require("./util/location");
const import_1 = require("./module/import");
const eps_1 = require("./rest/eps");
const cache = require("@midwayjs/cache");
let CoolConfiguration = class CoolConfiguration {
    async onReady(container) {
        // 处理模块配置
        await container.getAsync(config_1.CoolModuleConfig);
        // 导入模块数据
        await container.getAsync(import_1.CoolModuleImport);
        // 常用函数处理
        await container.getAsync(func_1.FuncUtil);
        // 常用函数处理
        await container.getAsync(data_1.CoolUrlTagData);
        // 事件
        await container.getAsync(index_1.CoolEventManager);
        if (this.app.getEnv() == 'local') {
            // 实体与路径
            const eps = await container.getAsync(eps_1.CoolEps);
            eps.init();
        }
        // 异常处理
        this.app.useFilter([filter_1.CoolExceptionFilter]);
        // 清除 location
        setTimeout(() => {
            location_1.default.clean();
            this.coreLogger.info('\x1B[36m [cool:core] location clean \x1B[0m');
        }, 10000);
    }
    async onServerReady() {
        location_1.default.clean();
    }
};
__decorate([
    decorator_1.App(),
    __metadata("design:type", Object)
], CoolConfiguration.prototype, "app", void 0);
__decorate([
    decorator_1.Logger(),
    __metadata("design:type", Object)
], CoolConfiguration.prototype, "coreLogger", void 0);
__decorate([
    decorator_1.Inject(),
    __metadata("design:type", koa.Framework)
], CoolConfiguration.prototype, "framework", void 0);
CoolConfiguration = __decorate([
    decorator_1.Configuration({
        namespace: 'cool',
        imports: [cache],
        importConfigs: [
            {
                default: DefaultConfig,
            },
        ],
    })
], CoolConfiguration);
exports.CoolConfiguration = CoolConfiguration;
