// @midwayjs/mwcc/dist/transformation/null-transformation-context.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nullTransformationContext = void 0;
const typescript_1 = __importDefault(require("typescript"));
exports.nullTransformationContext = {
    factory: typescript_1.default.factory,
    enableEmitNotification: noop,
    enableSubstitution: noop,
    endLexicalEnvironment: returnUndefined,
    getCompilerOptions: () => ({}),
    getEmitHost: notImplemented,
    getEmitResolver: notImplemented,
    setLexicalEnvironmentFlags: noop,
    getLexicalEnvironmentFlags: () => 0,
    hoistFunctionDeclaration: noop,
    hoistVariableDeclaration: noop,
    addInitializationStatement: noop,
    isEmitNotificationEnabled: notImplemented,
    isSubstitutionEnabled: notImplemented,
    onEmitNode: noop,
    onSubstituteNode: notImplemented,
    readEmitHelpers: notImplemented,
    requestEmitHelper: noop,
    resumeLexicalEnvironment: noop,
    startLexicalEnvironment: noop,
    suspendLexicalEnvironment: noop,
    addDiagnostic: noop,
};
function noop(_) { }
function returnUndefined() {
    return undefined;
}
function notImplemented() {
    throw new Error('Not implemented');
}
