// @midwayjs/core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MidwayFrameworkType = exports.MidwayMockService = exports.MidwayDecoratorService = exports.MidwayMiddlewareService = exports.MidwayLifeCycleService = exports.MidwayAspectService = exports.MidwayFrameworkService = exports.MidwayLoggerService = exports.MidwayInformationService = exports.MidwayEnvironmentService = exports.MidwayConfigService = exports.createConfiguration = exports.extend = exports.wrapMiddleware = exports.pathMatching = exports.transformRequestObjectByType = exports.deprecatedOutput = exports.delegateTargetAllPrototypeMethod = exports.delegateTargetProperties = exports.delegateTargetMethod = exports.delegateTargetPrototypeMethod = exports.safeRequire = exports.safelyGet = exports.BaseFramework = exports.MidwayRequestContainer = void 0;
__exportStar(require("./interface"), exports);
__exportStar(require("./context/container"), exports);
var requestContainer_1 = require("./context/requestContainer");
Object.defineProperty(exports, "MidwayRequestContainer", { enumerable: true, get: function () { return requestContainer_1.MidwayRequestContainer; } });
var baseFramework_1 = require("./baseFramework");
Object.defineProperty(exports, "BaseFramework", { enumerable: true, get: function () { return baseFramework_1.BaseFramework; } });
__exportStar(require("./context/providerWrapper"), exports);
__exportStar(require("./common/constants"), exports);
var util_1 = require("./util/");
Object.defineProperty(exports, "safelyGet", { enumerable: true, get: function () { return util_1.safelyGet; } });
Object.defineProperty(exports, "safeRequire", { enumerable: true, get: function () { return util_1.safeRequire; } });
Object.defineProperty(exports, "delegateTargetPrototypeMethod", { enumerable: true, get: function () { return util_1.delegateTargetPrototypeMethod; } });
Object.defineProperty(exports, "delegateTargetMethod", { enumerable: true, get: function () { return util_1.delegateTargetMethod; } });
Object.defineProperty(exports, "delegateTargetProperties", { enumerable: true, get: function () { return util_1.delegateTargetProperties; } });
Object.defineProperty(exports, "delegateTargetAllPrototypeMethod", { enumerable: true, get: function () { return util_1.delegateTargetAllPrototypeMethod; } });
Object.defineProperty(exports, "deprecatedOutput", { enumerable: true, get: function () { return util_1.deprecatedOutput; } });
Object.defineProperty(exports, "transformRequestObjectByType", { enumerable: true, get: function () { return util_1.transformRequestObjectByType; } });
Object.defineProperty(exports, "pathMatching", { enumerable: true, get: function () { return util_1.pathMatching; } });
Object.defineProperty(exports, "wrapMiddleware", { enumerable: true, get: function () { return util_1.wrapMiddleware; } });
var extend_1 = require("./util/extend");
Object.defineProperty(exports, "extend", { enumerable: true, get: function () { return extend_1.extend; } });
__exportStar(require("./util/pathFileUtil"), exports);
__exportStar(require("./util/webRouterParam"), exports);
__exportStar(require("./common/webRouterCollector"), exports);
__exportStar(require("./common/triggerCollector"), exports);
var configuration_1 = require("./functional/configuration");
Object.defineProperty(exports, "createConfiguration", { enumerable: true, get: function () { return configuration_1.createConfiguration; } });
var configService_1 = require("./service/configService");
Object.defineProperty(exports, "MidwayConfigService", { enumerable: true, get: function () { return configService_1.MidwayConfigService; } });
var environmentService_1 = require("./service/environmentService");
Object.defineProperty(exports, "MidwayEnvironmentService", { enumerable: true, get: function () { return environmentService_1.MidwayEnvironmentService; } });
var informationService_1 = require("./service/informationService");
Object.defineProperty(exports, "MidwayInformationService", { enumerable: true, get: function () { return informationService_1.MidwayInformationService; } });
var loggerService_1 = require("./service/loggerService");
Object.defineProperty(exports, "MidwayLoggerService", { enumerable: true, get: function () { return loggerService_1.MidwayLoggerService; } });
var frameworkService_1 = require("./service/frameworkService");
Object.defineProperty(exports, "MidwayFrameworkService", { enumerable: true, get: function () { return frameworkService_1.MidwayFrameworkService; } });
var aspectService_1 = require("./service/aspectService");
Object.defineProperty(exports, "MidwayAspectService", { enumerable: true, get: function () { return aspectService_1.MidwayAspectService; } });
var lifeCycleService_1 = require("./service/lifeCycleService");
Object.defineProperty(exports, "MidwayLifeCycleService", { enumerable: true, get: function () { return lifeCycleService_1.MidwayLifeCycleService; } });
var middlewareService_1 = require("./service/middlewareService");
Object.defineProperty(exports, "MidwayMiddlewareService", { enumerable: true, get: function () { return middlewareService_1.MidwayMiddlewareService; } });
var decoratorService_1 = require("./service/decoratorService");
Object.defineProperty(exports, "MidwayDecoratorService", { enumerable: true, get: function () { return decoratorService_1.MidwayDecoratorService; } });
var mockService_1 = require("./service/mockService");
Object.defineProperty(exports, "MidwayMockService", { enumerable: true, get: function () { return mockService_1.MidwayMockService; } });
__exportStar(require("./service/pipelineService"), exports);
__exportStar(require("./util/contextUtil"), exports);
__exportStar(require("./common/serviceFactory"), exports);
__exportStar(require("./common/dataListener"), exports);
__exportStar(require("./common/fileDetector"), exports);
__exportStar(require("./common/webGenerator"), exports);
__exportStar(require("./common/middlewareManager"), exports);
__exportStar(require("./util/pathToRegexp"), exports);
__exportStar(require("./util/httpclient"), exports);
__exportStar(require("./common/filterManager"), exports);
__exportStar(require("./common/applicationManager"), exports);
__exportStar(require("./setup"), exports);
__exportStar(require("./error"), exports);
/**
 * proxy
 */
var decorator_1 = require("@midwayjs/decorator");
Object.defineProperty(exports, "MidwayFrameworkType", { enumerable: true, get: function () { return decorator_1.MidwayFrameworkType; } });
