// @midwayjs/core/dist/interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MIDWAY_LOGGER_WRITEABLE_DIR = exports.MidwayProcessTypeEnum = exports.REQUEST_CTX_LOGGER_CACHE_KEY = exports.HTTP_SERVER_KEY = exports.REQUEST_OBJ_CTX_KEY = exports.REQUEST_CTX_KEY = exports.ObjectLifeCycleEvent = void 0;
var ObjectLifeCycleEvent;
(function (ObjectLifeCycleEvent) {
    ObjectLifeCycleEvent["BEFORE_BIND"] = "beforeBind";
    ObjectLifeCycleEvent["BEFORE_CREATED"] = "beforeObjectCreated";
    ObjectLifeCycleEvent["AFTER_CREATED"] = "afterObjectCreated";
    ObjectLifeCycleEvent["AFTER_INIT"] = "afterObjectInit";
    ObjectLifeCycleEvent["BEFORE_DESTROY"] = "beforeObjectDestroy";
})(ObjectLifeCycleEvent = exports.ObjectLifeCycleEvent || (exports.ObjectLifeCycleEvent = {}));
exports.REQUEST_CTX_KEY = 'ctx';
exports.REQUEST_OBJ_CTX_KEY = '_req_ctx';
exports.HTTP_SERVER_KEY = '_midway_http_server';
exports.REQUEST_CTX_LOGGER_CACHE_KEY = '_midway_ctx_logger_cache';
var MidwayProcessTypeEnum;
(function (MidwayProcessTypeEnum) {
    MidwayProcessTypeEnum["APPLICATION"] = "APPLICATION";
    MidwayProcessTypeEnum["AGENT"] = "AGENT";
})(MidwayProcessTypeEnum = exports.MidwayProcessTypeEnum || (exports.MidwayProcessTypeEnum = {}));
exports.MIDWAY_LOGGER_WRITEABLE_DIR = 'MIDWAY_LOGGER_WRITEABLE_DIR';
