// @midwayjs/core/dist/setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareGlobalApplicationContext = exports.destroyGlobalApplicationContext = exports.initializeGlobalApplicationContext = void 0;
const _1 = require("./");
const config_default_1 = require("./config/config.default");
const decorator_1 = require("@midwayjs/decorator");
const util = require("util");
const path_1 = require("path");
const logger_1 = require("@midwayjs/logger");
const debug = util.debuglog('midway:debug');
/**
 * midway framework main entry, this method bootstrap all service and framework.
 * @param globalOptions
 */
async function initializeGlobalApplicationContext(globalOptions) {
    const applicationContext = prepareGlobalApplicationContext(globalOptions);
    // init logger
    const loggerService = await applicationContext.getAsync(_1.MidwayLoggerService, [
        applicationContext,
    ]);
    if (loggerService.getLogger('appLogger')) {
        // register global logger
        applicationContext.registerObject('logger', loggerService.getLogger('appLogger'));
    }
    // framework/config/plugin/logger/app decorator support
    await applicationContext.getAsync(_1.MidwayFrameworkService, [
        applicationContext,
        globalOptions,
    ]);
    // lifecycle support
    await applicationContext.getAsync(_1.MidwayLifeCycleService, [
        applicationContext,
    ]);
    // some preload module init
    const modules = (0, decorator_1.listPreloadModule)();
    for (const module of modules) {
        // preload init context
        await applicationContext.getAsync(module);
    }
    return applicationContext;
}
exports.initializeGlobalApplicationContext = initializeGlobalApplicationContext;
async function destroyGlobalApplicationContext(applicationContext) {
    // stop lifecycle
    const lifecycleService = await applicationContext.getAsync(_1.MidwayLifeCycleService);
    await lifecycleService.stop();
    // stop container
    await applicationContext.stop();
    (0, decorator_1.clearBindContainer)();
    logger_1.loggers.close();
    global['MIDWAY_APPLICATION_CONTEXT'] = undefined;
    global['MIDWAY_MAIN_FRAMEWORK'] = undefined;
}
exports.destroyGlobalApplicationContext = destroyGlobalApplicationContext;
/**
 * prepare applicationContext, it use in egg framework.
 * @param globalOptions
 */
function prepareGlobalApplicationContext(globalOptions) {
    var _a, _b, _c, _d;
    debug('[core]: start "initializeGlobalApplicationContext"');
    debug(`[core]: bootstrap options = ${util.inspect(globalOptions)}`);
    const appDir = (_a = globalOptions.appDir) !== null && _a !== void 0 ? _a : '';
    const baseDir = (_b = globalOptions.baseDir) !== null && _b !== void 0 ? _b : '';
    // new container
    const applicationContext = (_c = globalOptions.applicationContext) !== null && _c !== void 0 ? _c : new _1.MidwayContainer();
    // bind container to decoratorManager
    debug('[core]: delegate module map from decoratorManager');
    (0, decorator_1.bindContainer)(applicationContext);
    global['MIDWAY_APPLICATION_CONTEXT'] = applicationContext;
    // register baseDir and appDir
    applicationContext.registerObject('baseDir', baseDir);
    applicationContext.registerObject('appDir', appDir);
    if (globalOptions.moduleDetector !== false) {
        if (globalOptions.moduleDetector === undefined ||
            globalOptions.moduleDetector === 'file') {
            applicationContext.setFileDetector(new _1.DirectoryFileDetector({
                loadDir: baseDir,
                ignore: (_d = globalOptions.ignore) !== null && _d !== void 0 ? _d : [],
            }));
        }
        else if (globalOptions.moduleDetector) {
            applicationContext.setFileDetector(globalOptions.moduleDetector);
        }
    }
    // bind inner service
    applicationContext.bindClass(_1.MidwayEnvironmentService);
    applicationContext.bindClass(_1.MidwayInformationService);
    applicationContext.bindClass(_1.MidwayAspectService);
    applicationContext.bindClass(_1.MidwayDecoratorService);
    applicationContext.bindClass(_1.MidwayConfigService);
    applicationContext.bindClass(_1.MidwayLoggerService);
    applicationContext.bindClass(_1.MidwayApplicationManager);
    applicationContext.bindClass(_1.MidwayFrameworkService);
    applicationContext.bindClass(_1.MidwayMiddlewareService);
    applicationContext.bindClass(_1.MidwayLifeCycleService);
    applicationContext.bindClass(_1.MidwayMockService);
    // bind preload module
    if (globalOptions.preloadModules && globalOptions.preloadModules.length) {
        for (const preloadModule of globalOptions.preloadModules) {
            applicationContext.bindClass(preloadModule);
        }
    }
    // init default config
    const configService = applicationContext.get(_1.MidwayConfigService);
    configService.add([
        {
            default: config_default_1.default,
        },
    ]);
    // init aop support
    applicationContext.get(_1.MidwayAspectService, [applicationContext]);
    // init decorator service
    applicationContext.get(_1.MidwayDecoratorService, [applicationContext]);
    if (!globalOptions.imports) {
        globalOptions.imports = [
            (0, _1.safeRequire)((0, path_1.join)(globalOptions.baseDir, 'configuration')),
        ];
    }
    for (const configurationModule of []
        .concat(globalOptions.imports)
        .concat(globalOptions.configurationModule)) {
        // load configuration and component
        if (configurationModule) {
            applicationContext.load(configurationModule);
        }
    }
    // bind user code module
    applicationContext.ready();
    if (globalOptions.globalConfig) {
        if (Array.isArray(globalOptions.globalConfig)) {
            configService.add(globalOptions.globalConfig);
        }
        else {
            configService.addObject(globalOptions.globalConfig);
        }
    }
    // merge config
    configService.load();
    debug('[core]: Current config = %j', configService.getConfiguration());
    // middleware support
    applicationContext.get(_1.MidwayMiddlewareService, [applicationContext]);
    return applicationContext;
}
exports.prepareGlobalApplicationContext = prepareGlobalApplicationContext;
