// @midwayjs/decorator/dist/decorator/web/paramMapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fields = exports.Queries = exports.RequestIP = exports.RequestPath = exports.Files = exports.File = exports.Headers = exports.Param = exports.Query = exports.Body = exports.Session = exports.RouteParamTypes = void 0;
const __1 = require("../../");
var RouteParamTypes;
(function (RouteParamTypes) {
    RouteParamTypes[RouteParamTypes["QUERY"] = 0] = "QUERY";
    RouteParamTypes[RouteParamTypes["BODY"] = 1] = "BODY";
    RouteParamTypes[RouteParamTypes["PARAM"] = 2] = "PARAM";
    RouteParamTypes[RouteParamTypes["HEADERS"] = 3] = "HEADERS";
    RouteParamTypes[RouteParamTypes["SESSION"] = 4] = "SESSION";
    RouteParamTypes[RouteParamTypes["FILESTREAM"] = 5] = "FILESTREAM";
    RouteParamTypes[RouteParamTypes["FILESSTREAM"] = 6] = "FILESSTREAM";
    RouteParamTypes[RouteParamTypes["NEXT"] = 7] = "NEXT";
    RouteParamTypes[RouteParamTypes["REQUEST_PATH"] = 8] = "REQUEST_PATH";
    RouteParamTypes[RouteParamTypes["REQUEST_IP"] = 9] = "REQUEST_IP";
    RouteParamTypes[RouteParamTypes["QUERIES"] = 10] = "QUERIES";
    RouteParamTypes[RouteParamTypes["FIELDS"] = 11] = "FIELDS";
})(RouteParamTypes = exports.RouteParamTypes || (exports.RouteParamTypes = {}));
const createParamMapping = function (type) {
    return (propertyData) => {
        return (0, __1.createCustomParamDecorator)(__1.WEB_ROUTER_PARAM_KEY, {
            type,
            propertyData,
        });
    };
};
const Session = (property) => createParamMapping(RouteParamTypes.SESSION)(property);
exports.Session = Session;
const Body = (property) => createParamMapping(RouteParamTypes.BODY)(property);
exports.Body = Body;
const Query = (property) => createParamMapping(RouteParamTypes.QUERY)(property);
exports.Query = Query;
const Param = (property) => createParamMapping(RouteParamTypes.PARAM)(property);
exports.Param = Param;
const Headers = (property) => createParamMapping(RouteParamTypes.HEADERS)(property);
exports.Headers = Headers;
const File = (property) => createParamMapping(RouteParamTypes.FILESTREAM)(property);
exports.File = File;
const Files = (property) => createParamMapping(RouteParamTypes.FILESSTREAM)(property);
exports.Files = Files;
const RequestPath = () => createParamMapping(RouteParamTypes.REQUEST_PATH)();
exports.RequestPath = RequestPath;
const RequestIP = () => createParamMapping(RouteParamTypes.REQUEST_IP)();
exports.RequestIP = RequestIP;
const Queries = (property) => createParamMapping(RouteParamTypes.QUERIES)(property);
exports.Queries = Queries;
const Fields = (property) => createParamMapping(RouteParamTypes.FIELDS)(property);
exports.Fields = Fields;
